### PR TITLE
webuipoc: Restructure for JS framework PoCs

### PR DESCRIPTION
**Pull request by ricekot**
Update the add-on build script to build and package PoCs that use JS frameworks like React and follow a few set guidelines.

Include a very minimal React example to demonstrate and test the Gradle task.

The react app was generated with [nano-react-app](https://github.com/nano-react-app/nano-react-app) and slightly modified.

### CODE REVIEW

File: addOns/webuipoc/README.md [hunk 1]
```diff
@@ -35,16 +35,22 @@ We would still like you to contribute a basic PoC but we are very happy for you
...
 You may well find it easier to edit these files in the 'webuipoc' directory used by the ZAP runtime, this will be under
 the ZAP [default directory](https://www.zaproxy.org/faq/what-is-the-default-directory-that-zap-uses/).
-Once everything is working how you want you can copy them back to your new directory under 
-[webuipoc](src/main/zapHomeFiles/webuipoc) for your pull request.
-
-Source TypeScript files should be added to a new directory under the [src](src) directory.
+Once everything is working how you want you can copy them back to your new directory under
+[src/main/pocs](src/main/pocs) for your pull request.
+
+If your PoC uses React or some other JS framework, then it must follow the below guidelines so that the build files
```
Comment (kingthorin) [inline addOns/webuipoc/README.md:47]: Swap the order on "below guidelines" ➡️ "guidelines below"

Comment (kingthorin) [reply addOns/webuipoc/README.md:47]: That's the only thing I see but you and @thc202 probably know/understand the build better than I do.

**Comment by njmulsqb** [comment]
Thanks for the PR, is there any special reason for preferring nano-react-app over create-react-app which is more commonly used?

**Comment by kingthorin** [comment]
I believe simply to keep things small for the example

File: addOns/webuipoc/README.md [hunk 2]
```diff
@@ -35,16 +35,22 @@ We would still like you to contribute a basic PoC but we are very happy for you
...
-Once everything is working how you want you can copy them back to your new directory under 
-[webuipoc](src/main/zapHomeFiles/webuipoc) for your pull request.
-
-Source TypeScript files should be added to a new directory under the [src](src) directory.
+Once everything is working how you want you can copy them back to your new directory under
+[src/main/pocs](src/main/pocs) for your pull request.
+
+If your PoC uses React or some other JS framework, then it must follow the guidelines below so that the build files
+are automatically packaged with the add-on when it is built:
+- Its dependencies can be installed by running `npm install` in the root directory of the add-on.
```
Comment (thc202) [inline addOns/webuipoc/README.md:49]: `of the PoC.`? (same for following two lines)

File: addOns/webuipoc/webuipoc.gradle.kts [hunk 3]
```diff
@@ -1,5 +1,62 @@
...
+
+val pocsSrcDir = projectDir.resolve("src/main/pocs")
+val pocsBuildDir = layout.buildDirectory.dir("pocs").get()
+val pocBuildTasks: MutableList<TaskProvider<*>> = mutableListOf()
+for (dir in pocsSrcDir.listFiles()!!) {
+    val packageJson = File(dir, "package.json")
+    val outputDir = pocsBuildDir.dir("webuipoc").dir(dir.name)
+    if (packageJson.exists()) {
+        val installTask =
+            tasks.register<NpmTask>("installPoc${dir.name.replaceFirstChar { it.titlecase(Locale.getDefault()) }}Dependencies") {
```
[the remark this answers is not in the document]
Comment (ricekot) [reply addOns/webuipoc/webuipoc.gradle.kts:23]: That seems to have been deprecated. Will use `org.gradle.configurationcache.extensions.capitalized` instead.

[the remark this answers is not in the document]
Comment (thc202) [reply addOns/webuipoc/webuipoc.gradle.kts]: Sorry, should have checked its state.

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]

**Pull request merged**